<div class="alert alert-success">
<b>Комментарий ревьюера v 3.0 ✔️:</b> Отлично, Дмитрий! Ты исправил все, даже самые незначительные недостатки и теперь твой проект конечно же принят.
    
Спасибо за работу и успехов в следующих проектах!)

<div class="alert alert-danger">
<b>Комментарий ревьюера v 2.0 ❌:</b> Дмитрий, ты очень хорошо поработал над комментариями, но все же у меня осталось еще одно замечание, оно отмечено v 2.0 и красным цветом. Пожалуйста, поправь его и я буду рад принять твой проект)

<div style="border:solid green 2px; padding: 20px">
<b>Привет!</b>

Меня зовут Михаленко Дмитрий, и я буду проверять твой проект. Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так выделены небольшие замечания. Я надеюсь, что их ты тоже учтешь - твой проект от этого станет только лучше. Но настаивать на их отработке не буду.

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом.** Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее.

# Исследование надежности заемщиков.

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Хорошим тоном считается, когда перед проведением исследования обозначаются его цели и задачи)

<div class="alert alert-info"> <b>Комментарий студента:</b>
Цель проекта - определить зависимость между семейным положением и выплатой в срок кредита клиента, а также между количеством детей и клиента и  выплатой в срок кредита.
Задачи:
1) Ознакомиться с данными. Выявить баги.
2) Выполнить предобратоку данных. 
3) Проверка гипотез.
4) Выводы на освнове результатов проверки гипотез.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Теперь все как надо

### Шаг 1. Обзор данных

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Заметь, ячейки в тетрадке исполняются, начиная не с единицы. Перед отправкой проектов на ревью советую нажимать на панели управления Kernel - Restart & Run All

In [ ]:
import pandas as pd
data=pd.read_csv('/datasets/data.csv')

In [ ]:
print(data)

       children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем        

In [ ]:
print(data.columns)

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')


In [ ]:
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-info"> <b>Комментарий студента:</b> Выводы исходя из полученных данных: всего 12 столбцов, 21525 строк. В столбцах имеются целочисленные значения, десятичные и объект(строка). В столбцах days_employed и total_income есть пропуски. Также можно обратить внимание на отрицательное кол-во дней в столбце days_employed и десятичное значение дней.</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Хорошо делаешь выводы по данным, но все же их стоит офрмлять в отдельных ячейках, так твои пояснения будут более "читаемыми"

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Здорово

### Шаг 2.1 Заполнение пропусков

In [ ]:
display(data[data['total_income'].isna()])

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

<div class="alert alert-info"> <b>Комментарий студента:</b> 
    Таким образом, мы видим, что в стобцах days_employed и total_income есть пропущенные значения NaN. Такие пропуски принадлежат к типу данных float, то есть с ними можно проводить арифметические операции. Однако такие данные, как количество дней проработанных на работе и суммарный доход могут напрямую влиять на анализ, то лучше пропуски не удалять, а заменить их медианными значениями. </div>

In [ ]:
nan_part_in_total_income = data['total_income'].isna().sum()/len(data['total_income'])

In [ ]:
print('Доля пропусков в столбце total_income', nan_part_in_total_income)

Доля пропусков в столбце total_income 0.10099883855981417


In [ ]:
nan_part_in_days_employed = data['days_employed'].isna().sum()/len(data['days_employed'])

In [ ]:
print('Доля пропусков в столбце days_employed', nan_part_in_days_employed)

Доля пропусков в столбце days_employed 0.10099883855981417


<div class="alert alert-info"> <b>Комментарий студента:</b> 
    Доля пропусков для двух столбцов одинаковая. Пенсионеры, госслужащие получают точно официально заработную плату|пенсию. Поэтому данная ошибка больше похожа на механическую ошибку при получении данных, нежели на отсутствие официальной работы или отсутствие какого-либо дохода. Т.к. исходя из этого можно подумать, что пропуски не являются отсутсвием дохода у людей, а также работы, нельзя просто заменить значения на 0. При этом удалить строки с пропусками тоже нельзя, так как это окажет влияние на исследование, так как пропуски соствляют около 10% от всего объема. В таком случае лучше заменить пропуски медианным значением, что не окажет сильного влияния на итоги исследования.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Все отлично

In [ ]:
data['total_income']=data['total_income'].fillna(data['total_income'].median())

In [ ]:
display(data['total_income'].head(25))

0     253875.639453
1     112080.014102
2     145885.952297
3     267628.550329
4     158616.077870
5     255763.565419
6     240525.971920
7     135823.934197
8      95856.832424
9     144425.938277
10    113943.491460
11     77069.234271
12    145017.937533
13    130458.228857
14    165127.911772
15    116820.904450
16    289202.704229
17     90410.586745
18     56823.777243
19    242831.107982
20    209552.644578
21    131812.203909
22    149681.140731
23    128265.720871
24    290547.235997
Name: total_income, dtype: float64

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Написано все верно, но ты заполняешь данные средним(mean), а не медианой(median). Пожалуйста обрати на это внимание

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Хорошо, что поправил

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [ ]:
data['days_employed'] = abs(data['days_employed']) # берем модуль

In [ ]:
display(data['days_employed'])

0          8437.673028
1          4024.803754
2          5623.422610
3          4124.747207
4        340266.072047
             ...      
21520      4529.316663
21521    343937.404131
21522      2113.346888
21523      3112.481705
21524      1984.507589
Name: days_employed, Length: 21525, dtype: float64

In [ ]:
print(data['days_employed'].mean(), data['days_employed'].max(), data['days_employed'].min())

66914.72890682236 401755.40047533 24.14163324048118


In [ ]:
def hours_to_days(days_employed):
    if days_employed > 300000:
        days_employed = days_employed/24
        return days_employed
    else:
        return days_employed


In [ ]:
data['days_employed'] = data['days_employed'].apply(hours_to_days) #исходя из расчетов, посчитал, что такие большие значения это часы, заменил часы на дни

In [ ]:
print(data['days_employed'])

0         8437.673028
1         4024.803754
2         5623.422610
3         4124.747207
4        14177.753002
             ...     
21520     4529.316663
21521    14330.725172
21522     2113.346888
21523     3112.481705
21524     1984.507589
Name: days_employed, Length: 21525, dtype: float64


In [ ]:
data['days_employed']=data['days_employed'].fillna(data['days_employed'].median()) #заменяем пропуски медианным способом

In [ ]:
display(data['days_employed'].head(25))
#постарался понять, что лучше исправить, но ощущение, что не все нашел, не все понял, если будут замечания по этому пункту, то буду рад)

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4     14177.753002
5       926.185831
6      2879.202052
7       152.779569
8      6929.865299
9      2188.756445
10     4171.483647
11      792.701887
12     2194.220567
13     1846.641941
14     1844.956182
15      972.364419
16     1719.934226
17     2369.999720
18    16678.380705
19    10038.818549
20     1311.604166
21      253.685166
22     1766.644138
23      272.981385
24    14106.331371
Name: days_employed, dtype: float64

In [ ]:
print(data['children'].describe()) #странно, что есть минус и максимально 20 детей.

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64


In [ ]:
print(data.groupby('children').count())

          days_employed  dob_years  education  education_id  family_status  \
children                                                                     
-1                   47         47         47            47             47   
 0                14149      14149      14149         14149          14149   
 1                 4818       4818       4818          4818           4818   
 2                 2055       2055       2055          2055           2055   
 3                  330        330        330           330            330   
 4                   41         41         41            41             41   
 5                    9          9          9             9              9   
 20                  76         76         76            76             76   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
-1                      47      47           47     47     

<div class="alert alert-info"> <b>Комментарий студента:</b> Количество значений -1 и 20 не такое больше по сравнению со всей таблицей, такие строки можно удалить, либо -1 можно заменить на модуль, т.к. скорее всего это просто опечатка, а строки со значенями 20 детей можно удалить, т.к. поинтересоваться откуда такое значение появилось не у кого. </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Верно

In [ ]:
data['children']=abs(data['children']) #взял модуль

In [ ]:
data = data.loc[data['children'] < 20] #отфильтровал строки, чтобы исключить ззначение 20 для детей

In [ ]:
print(data.groupby('children').count())

          days_employed  dob_years  education  education_id  family_status  \
children                                                                     
0                 14149      14149      14149         14149          14149   
1                  4865       4865       4865          4865           4865   
2                  2055       2055       2055          2055           2055   
3                   330        330        330           330            330   
4                    41         41         41            41             41   
5                     9          9          9             9              9   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
0                    14149   14149        14149  14149         14149    14149  
1                     4865    4865         4865   4865          4865     4865  
2                     2055    2055         2055   2055 

In [ ]:
print(data['dob_years'].describe())

count    21449.000000
mean        43.298615
std         12.576314
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64


In [ ]:
print(data.groupby('dob_years').count()) #странно, что есть 0 в годах клиента

           children  days_employed  education  education_id  family_status  \
dob_years                                                                    
0               100            100        100           100            100   
19               14             14         14            14             14   
20               51             51         51            51             51   
21              110            110        110           110            110   
22              183            183        183           183            183   
23              253            253        253           253            253   
24              263            263        263           263            263   
25              356            356        356           356            356   
26              407            407        407           407            407   
27              491            491        491           491            491   
28              503            503        503           503     

In [ ]:
data = data.loc[data['dob_years'] > 0]

In [ ]:
print(data.groupby('dob_years').count())

           children  days_employed  education  education_id  family_status  \
dob_years                                                                    
19               14             14         14            14             14   
20               51             51         51            51             51   
21              110            110        110           110            110   
22              183            183        183           183            183   
23              253            253        253           253            253   
24              263            263        263           263            263   
25              356            356        356           356            356   
26              407            407        407           407            407   
27              491            491        491           491            491   
28              503            503        503           503            503   
29              543            543        543           543     

In [ ]:
print(data.groupby('education').count()) # разная высота букв, но будем делать работу с этим столбцом позже

                     children  days_employed  dob_years  education_id  \
education                                                               
ВЫСШЕЕ                    271            271        271           271   
Высшее                    266            266        266           266   
НАЧАЛЬНОЕ                  17             17         17            17   
НЕОКОНЧЕННОЕ ВЫСШЕЕ        29             29         29            29   
Начальное                  15             15         15            15   
Неоконченное высшее        47             47         47            47   
СРЕДНЕЕ                   768            768        768           768   
Среднее                   702            702        702           702   
УЧЕНАЯ СТЕПЕНЬ              1              1          1             1   
Ученая степень              1              1          1             1   
высшее                   4674           4674       4674          4674   
начальное                 250            250       

In [ ]:
print(data.groupby('gender').count()) # пол XNA, одно значение, возможно это такая идентификация пола, не буду удалять, на результат не повлияет

        children  days_employed  dob_years  education  education_id  \
gender                                                                
F          14118          14118      14118      14118         14118   
M           7230           7230       7230       7230          7230   
XNA            1              1          1          1             1   

        family_status  family_status_id  income_type   debt  total_income  \
gender                                                                      
F               14118             14118        14118  14118         14118   
M                7230              7230         7230   7230          7230   
XNA                 1                 1            1      1             1   

        purpose  
gender           
F         14118  
M          7230  
XNA           1  


<div class="alert alert-info"> <b>Комментарий студента:</b> В целом, я проверил все столбцы, больше ничего интересного не нашел, либо нашел, но в следующих пунктах будет вестись работа над данными в этих столбцах. </div>

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Также замечание касательно среднего(mean) и медианы(median). Замена часов на дни хорошая. Однако, тут стоит еще посмотреть на другие данные, там тоже могут быть аномалии

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Да, согласен

### Шаг 2.3. Изменение типов данных.

In [ ]:
data['total_income'] = data['total_income'].astype('int')
display(data['total_income'])

0        253875
1        112080
2        145885
3        267628
4        158616
          ...  
21520    224791
21521    155999
21522     89672
21523    244093
21524     82047
Name: total_income, Length: 21349, dtype: int64

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Здорово, что выбрал оптимальный тип данных
</div>



### Шаг 2.4. Удаление дубликатов.

In [ ]:
print('Дубликатов в таблице:', data.duplicated().sum()) #проверил общее количество дубликатов в таблице, дубликатов мало, поэтому можно применить ко всей таблице drop_duplicates()

Дубликатов в таблице: 54


In [ ]:
data=data.drop_duplicates().reset_index(drop=True)

In [ ]:
print('Дубликатов в таблице:', data.duplicated().sum())

Дубликатов в таблице: 0


In [ ]:
data=data.drop_duplicates().reset_index(drop=True)

In [ ]:
print('Дубликатов в таблице:', data.duplicated().sum())

Дубликатов в таблице: 0


In [ ]:
data['education'] = data['education'].str.lower() #привел к общему регистру столбец, но повторы в данном случае логичны, у всех будет с

In [ ]:
print(data['education'].value_counts()) #регистр в порядке, повторы логичны, их не надо менять/удалять

среднее                15065
высшее                  5202
неоконченное высшее      740
начальное                282
ученая степень             6
Name: education, dtype: int64


In [ ]:
print(data['family_status'].value_counts()) #проверил повторы и регистр

женат / замужем          12247
гражданский брак          4130
Не женат / не замужем     2785
в разводе                 1183
вдовец / вдова             950
Name: family_status, dtype: int64


In [ ]:
data['family_status'] = data['family_status'].str.lower()

In [ ]:
print(data['income_type'].value_counts()) #проверил повторы и регистр, все в порядке, повторы в данном столбце логичны, регистр в порядке

сотрудник          10994
компаньон           5038
пенсионер           3808
госслужащий         1449
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64


In [ ]:
print(data['purpose'].value_counts()) #проверил повторы и регистр, регистр в порядке, но много одинаковых по сути целей, которые записаны по-разному, данная задача будет решаться в последнем пункте

свадьба                                   787
на проведение свадьбы                     765
сыграть свадьбу                           760
операции с недвижимостью                  671
покупка коммерческой недвижимости         656
покупка жилья для сдачи                   649
операции с коммерческой недвижимостью     644
операции с жильем                         642
покупка жилья                             637
покупка жилья для семьи                   636
жилье                                     636
недвижимость                              628
строительство собственной недвижимости    627
операции со своей недвижимостью           626
строительство жилой недвижимости          621
строительство недвижимости                619
покупка своего жилья                      619
покупка недвижимости                      616
ремонт жилью                              603
покупка жилой недвижимости                601
на покупку своего автомобиля              502
заняться высшим образованием      

In [ ]:
print('Дубликатов в таблице:', data.duplicated().sum())

Дубликатов в таблице: 17


In [ ]:
data=data.drop_duplicates().reset_index(drop=True)

In [ ]:
print('Дубликатов в таблице:', data.duplicated().sum())

Дубликатов в таблице: 0


<div class="alert alert-info"> <b>Комментарий студента V 2.0:</b> Проверил на наличие дубликатов, они правда появились, удалил и проверил еще раз </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v 3.0 ✔️:</b> Молодец!

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Хорошо, но, после удаления неявных дубликатов стоит еще раз проверить data.duplicated().sum()
    
Вдруг, при исправлении неявных дубликатов случайно был создан явный

<div class="alert alert-danger">
<b>Комментарий ревьюера v 2.0 ❌:</b> Видимо случайно забыл про этот комментарий, проверь пожалуйста на дубликаты в конце этого раздела

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [ ]:
education_list = data[['education', 'education_id']]

In [ ]:
education_list = education_list.drop_duplicates().reset_index(drop=True)

In [ ]:
display(education_list)

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [ ]:
family_status_list = data[['family_status', 'family_status_id']]

In [ ]:
family_status_list = family_status_list.drop_duplicates().reset_index(drop=True)

In [ ]:
display(family_status_list)

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

In [ ]:
data = data.drop(columns=['education', 'family_status'])
print(data)

       children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0   14177.753002         53             1                 1   
...         ...            ...        ...           ...               ...   
21273         1    4529.316663         43             1                 1   
21274         0   14330.725172         67             1                 0   
21275         1    2113.346888         38             1                 1   
21276         3    3112.481705         38             1                 0   
21277         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose 

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Нам нужен словарь, чтобы однозначно соотнести значение education/family_status и цифровой идентификатор, поэтому нужно удалить дубликаты из education_list / family_status_list
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Теперь все здорово

### Шаг 2.6. Категоризация дохода.

In [ ]:
def category_income(total_income):
    if total_income <= 30000:
        return 'E'
    elif total_income <= 50000:
        return 'D'
    elif total_income <= 200000:
        return 'C'
    elif total_income <= 1000000:
        return 'B'
    else:
        return 'A'

data['total_income_category'] = data['total_income'].apply(category_income)
display(data)

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0   14177.753002         53             1                 1   
...         ...            ...        ...           ...               ...   
21273         1    4529.316663         43             1                 1   
21274         0   14330.725172         67             1                 0   
21275         1    2113.346888         38             1                 1   
21276         3    3112.481705         38             1                 0   
21277         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21273      F   компаньон     0        224791             операции с жильем   
21274      F   пенсионер     0        155999          сделка с автомобилем   
21275      M   сотрудник     1         89672                  недвижимость   
21276      M   сотрудник     1        244093  на покупку своего автомобиля   
21277      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21273                     B  
21274                     C  
21275                     C  
21276                     B  
21277                     C  

[21278 rows x 11 columns]

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Здорово, что используешь метод apply

### Шаг 2.7. Категоризация целей кредита.

In [ ]:
def purpose_category(item):
    
    purposes = ['автомобили', 'автомобиль', 'высшее образование',

       'дополнительное образование', 'жилье',

       'заняться высшим образованием', 'заняться образованием',

       'на покупку автомобиля', 'на покупку подержанного автомобиля',

       'на покупку своего автомобиля', 'на проведение свадьбы',

       'недвижимость', 'образование', 'операции с жильем',

       'операции с коммерческой недвижимостью',

       'операции с недвижимостью', 'операции со своей недвижимостью',

       'покупка жилой недвижимости', 'покупка жилья',

       'покупка жилья для сдачи', 'покупка жилья для семьи',

       'покупка коммерческой недвижимости', 'покупка недвижимости',

       'покупка своего жилья', 'получение высшего образования',

       'получение дополнительного образования', 'получение образования',

       'приобретение автомобиля', 'профильное образование',

       'ремонт жилью', 'свадьба', 'свой автомобиль',

       'сделка с автомобилем', 'сделка с подержанным автомобилем',

       'строительство жилой недвижимости', 'строительство недвижимости',

       'строительство собственной недвижимости', 'сыграть свадьбу']



    if 'свадьб' in item:

        return 'проведение свадьбы'

    if 'образов' in item:

        return 'получение образования'

    if 'автомобил' in item:

        return 'операции с автомобилем'

    return 'операции с недвижимостью'

            

data['purpose_category'] = data['purpose'].apply(purpose_category) # добавляю столбец purpose_category 

display(data.head(10))



children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   14177.753002         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Все верно, но я бы рекомендовал использовать не 4, а 5 категорий (пятая - "нет категории"), хотя это и не обязательно, но так мы бы обезопасили себя от недоразумений в анализе)

### Ответы на вопросы.

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Честно сказать, не совсем корректные пивот в этом разделе работы. Давай ты попробуешь в каждом из ответов на вопросы вывести такой пивот, в котором будет три столбца: Количество кредитополучателей(числом), количество должников(числом) и доля должников(mean этих чисел). Так таблица будет нагляднее

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправил пивоты, заменил выводы в соответствии с новыми данными </div>

##### Вопрос 1-2:

In [ ]:
pivot = data.pivot_table(index=['family_status_id'], columns='debt', values='dob_years', aggfunc='count').reset_index() #делаем сводную таблицу
pivot['%'] = pivot[1]/(pivot[0]+pivot[1])*100
pivot = pivot.merge(family_status_list, on='family_status_id', how='left').reset_index(drop=True)
display(pivot)

family_status_id      0    1         %          family_status
0                 0  11318  924  7.547786        женат / замужем
1                 1   3735  383  9.300631       гражданский брак
2                 2    888   62  6.526316         вдовец / вдова
3                 3   1099   84  7.100592              в разводе
4                 4   2513  272  9.766607  не женат / не замужем

In [ ]:
pivot = data.pivot_table(index=['children'], columns='debt', values='dob_years', aggfunc='count').reset_index() #делаем сводную таблицу
pivot['%'] = pivot[1]/(pivot[0]+pivot[1])*100
display(pivot)

debt  children        0       1         %
0            0  12964.0  1058.0  7.545286
1            1   4397.0   442.0  9.134119
2            2   1845.0   194.0  9.514468
3            3    301.0    27.0  8.231707
4            4     37.0     4.0  9.756098
5            5      9.0     NaN       NaN

Судя по таблицам с увеличением кол-ва детей, количество должников, которые не возвращают кредит, растет. При этом люди, в гражданском браке или не женатые менее стабильно выплачивают кредит, а категории людей в статусе "вдовец/вдова", "женат/замужем", "в разводе" примерно в одинаковом процентном соотношении более стабильно и регулярно выплачивают свой кредит.

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Согласен, хороший вывод

##### Вопрос 3-4:

In [ ]:
pivot_income_goal = data.pivot_table(index=['total_income_category'], columns='debt', values='dob_years', aggfunc='count').reset_index() #делаем сводную таблицу
pivot_income_goal['%'] = pivot_income_goal[1]/(pivot_income_goal[0]+pivot_income_goal[1])*100
display(pivot_income_goal)

debt total_income_category      0     1         %
0                        A     23     2  8.000000
1                        B   4642   353  7.067067
2                        C  14541  1347  8.478097
3                        D    327    21  6.034483
4                        E     20     2  9.090909

In [ ]:
pivot_purpose = data.pivot_table(index=['purpose_category'], columns='debt', values='dob_years', aggfunc='count').reset_index() #делаем сводную таблицу
pivot_purpose['%'] = pivot_purpose[1]/(pivot_purpose[0]+pivot_purpose[1])*100
display(pivot_purpose)

debt          purpose_category     0    1         %
0       операции с автомобилем  3871  398  9.323026
1     операции с недвижимостью  9951  777  7.242729
2        получение образования  3611  369  9.271357
3           проведение свадьбы  2120  181  7.866145

В процентном соотношении получается для зависимости дохода и выплаты кредита есть одно НО для категории "А" и "Е", я считаю, что не совсем коррректно делать какие-либо выводы - людей очень мало по сравнению с основной массой. Поэтому считаю, что наиболее рискованно выдавать кредит категории людей с доходом "С" или от 50000 до 200000, наиболее стабильно возвращают долг людей с доходом категории "D", хотя это довольной низкий доход. Возможно, проблема не только в доходах, но и в размерах самих кредитов, которые могут одобрить разным группам. Поэтому и у категорий с большим доходом возникает больше проблем с возвратом долга.
Если берется кредит для целей "операции с автомобилем" или "получение образования", то есть больший риск, что кредит будет не выплачен. Кредит под операции с недвижимостью и проведение свадьбы вылачивается стабильнее.

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Также все как надо, очень подробно

## Общий вывод:

1. Чем больше детей, тем выше вероятность, что кредит выплачиваться не будет.
2. Люди входящие в категории "в гражданском браке", "не женат / не замужем" менее стабильно выплачивают кредит, чем люди из категорий "вдовец/вдова", "женат/замужем", "в разводе".
3. Если берется кредит для целей "операции с автомобилем" или "получение образования", то есть больший риск, что кредит будет не выплачен. Кредит под операции с недвижимостью и проведение свадьбы вылачивается стабильнее.
4. Для категории "А" и "Е" не коррректно делать какие-либо выводы - людей очень мало по сравнению с основной массой.
5. Наиболее рискованно выдавать кредит категории людей с доходом "С" или от 50000 до 200000, наиболее стабильно возвращают долг людей с доходом категории "D", хотя это довольной низкий доход.

<div class="alert alert-success">
<b>Комментарий ревьюера v 2.0 ✔️:</b> Молодец!

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Выводы также необходимо будет переписать по новым пивотам

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Дмитрий, проект хороший! Ты последовательно проводишь анализ и неплохо владеешь инструментами для проведения исследования. Но всё же в проекте есть несколько замечаний и я попрошу тебя исправить их, чтобы он стал еще лучше! </div>